core/src/main/scala/cats/Traverse.scala

```diff
@@ -284,4 +285,30 @@ object Traverse {
   @deprecated("Use cats.syntax object imports", "2.2.0")
   object nonInheritedOps extends ToTraverseOps
 
+  private[cats] def traverseDirectly[G[_], A, B](
+    fa: IterableOnce[A]
+  )(f: A => G[B])(implicit G: StackSafeMonad[G]): G[Vector[B]] = {
+    fa.iterator.foldLeft(G.pure(Vector.empty[B])) { case (accG, a) =>
```

Comment:
I think we can do even better here by using a `Vector.newBuilder`, we can do even even better by using a dedicated builder for the desired collection (e.g. `ArraySeq` or `List`) and we can do even even even better by providing a size hint to the builder.

Reply:
I might have misunderstood something, but it looks like that's how things were done before the commit [8cc742e](https://github.com/typelevel/cats/pull/4498/commits/8cc742e469f3718516cdb4fb5894aa3f29c737ef) was introduced, which added the immutable data structures. :slightly_smiling_face:

Reply:
You're right :) let's just say this PR has been open a while and I lost track of the old stuff 😇